src/python/pants/base/specs.py

```diff
@@ -197,17 +197,21 @@ def make_glob_patterns(self, address_mapper):
   SiblingAddresses: 1,
   AscendantAddresses: 2,
   DescendantAddresses: 3,
-  type(None): 99
 }
 
 
-def more_specific(spec1: Spec, spec2: Spec) -> Spec:
+def more_specific(spec1: Optional[Spec], spec2: Optional[Spec]) -> Spec:
   """Returns which of the two specs is more specific.
 
   This is useful when a target matches multiple specs, and we want to associate it with
   the "most specific" one, which will make the most intuitive sense to the user.
   """
   # Note that if either of spec1 or spec2 is None, the other will be returned.
+  if spec1 is None:
+    assert spec2 is not None
```

Comment:
This assert looks like a change in behavior: I think this would have resulted in the method returning None before?

Either way, an assert wouldn't be friendly in this case: should consider either raising or returning `Optional[Spec]`.

Reply:
Yeah, asserts don't fire in production code, so we shouldn't really use them except in tests.

Also, why this more verbose way of achieving the same result as before?

Reply:
I was trying to satisfy mypy while moving fast and breaking things! This can likely be reverted.